### ISSUE TITLE
Implementation of DaNCE

### ISSUE THREAD

**Issue by BrandonHanx** [Confusezius/ECCV2020_DiVA_MultiFeature_DML#3]
https://github.com/Confusezius/ECCV2020_DiVA_MultiFeature_DML/blob/73c10e07e957457982257a097139f5811fedd087/criteria/fast_moco.py#L102-L106

Hey, I have some confusions about the implementation of DaNCE：

1. line103： I think we use -maxlog here is for the stability. So I guess this max operation should be handled row by row like `torch.max(distance_weighting, dim=1, keepdim=True)`. Rather than getting the maximum from the whole matrix.
2. line 106:  I guess we are doing normalization here for each N-pair. This normalization should be done for the distances among each anchor (with number as batch_size) and all negative samples (with number as memory_size). So I guess this line should be like `distance_weighting/torch.sum(distance_weighting, dim=1, keepdim=True) `.

Please point me out if I were wrong.

**Comment by Confusezius** [comment]
Thanks for the helpful comments! In both cases, you are very much right!
For 1., it was a copy-paste error from the initial distance miner code not adjusted for batch inputs. However, both cases work for what we want to achieve (stability). 
For 2., it should be indeed a normalization over the rowwise sum to achieve normalization of the weighting between 0 and 1. Now, it's essentially just a rescaling across all anchor-queue distances which we essentially directly address by choice of temperature. So as it is now, it is effectively a non-normalized reweighting, which techincally also achieves what we are going  for :).

In both cases, it shouldn't make much of a difference, but I will do some runs with this errors fixed and provide a commit with a `fast_move_fixed.py`with the errors fixed. You can also do a pull request, whatever comes first ;).

**Comment by BrandonHanx** [comment]
Thanks for your reply.

I am curious about if we will see any performance improvement after this bug is fixed. 😄 
Will stay tuned for your running results.

**Comment by Confusezius** [comment]
Alright, all runs finished - as expected, performance isn't impacted much, although convergence speeds do increase a bit which is nice :). I've added a note on the readme and a respective flag to turn on the fixes in the `fast_moco.py` file (`--diva_fixed`).
Thanks again for noticing! I'll close this for now, but feel free to reopen if you have any additional notes or questions!
